### PR TITLE
Scenes: Library panels pane in panel edit

### LINKED ISSUE #82347: [ON HOLD] Library panels: Implement tab to select a library panel in panel edit mode, replacing the panel with the library panel.

### PR DESCRIPTION
**Pull request by mdvictor**
Adds the Library pane to panel edit in scenes.

Closes https://github.com/grafana/grafana/issues/82347

### CODE REVIEW

File: public/app/features/dashboard-scene/panel-edit/PanelVizTypePicker.tsx [hunk 1]
```diff
@@ -21,25 +27,27 @@ export interface Props {
...
+  const { panel, sourcePanel } = vizManager.useState();
   const styles = useStyles2(getStyles);
   const [searchQuery, setSearchQuery] = useState('');
 
   const isWidgetEnabled = false;
   const tabKey = isWidgetEnabled ? LS_WIDGET_SELECT_TAB_KEY : LS_VISUALIZATION_SELECT_TAB_KEY;
   const defaultTab = isWidgetEnabled ? VisualizationSelectPaneTab.Widgets : VisualizationSelectPaneTab.Visualizations;
   const panelModel = useMemo(() => new PanelModelCompatibilityWrapper(panel), [panel]);
+  const vizPanel = sourcePanel.resolve();
+  const oldPanelModel = getOldPanelModel(vizPanel);
```
Comment (kaydelaney) [inline public/app/features/dashboard-scene/panel-edit/PanelVizTypePicker.tsx:39]: Can we memoize this so it's not recomputed for each render?

Comment (dprokop) [inline public/app/features/dashboard-scene/panel-edit/PanelVizTypePicker.tsx:39]: @mdvictor - we should not use the legacy model if possible. I suggest using `PanelModelCompatibilityWrapper` which is a PanelModel API around a `VizPanel`. Here's an example how to use it: https://github.com/grafana/grafana/blob/main/public/app/features/dashboard-scene/panel-edit/PanelVizTypePicker.tsx#L31

File: public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx [hunk 2]
```diff
@@ -409,6 +417,92 @@ export class VizPanelManager extends SceneObjectBase<VizPanelManagerState> {
...
+      title: panel.state.title,
+      uid: libPanel.uid,
+      name: libPanel.name,
+      panelKey: panel.state.key!,
+    });
+
+    libVizPanel.activate();
+
+    // we need to update panel edit and dashboard with the new library panel after it's been loaded
+    libVizPanel.subscribeToState(() => this.updatePanelFromLibPanel(libPanel, libVizPanel, vizPanel));
```
Comment (dprokop) [inline public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx:435]: Do we need to update the dashboard? I mean - shouldn't this happen when the changes are commited? In the old architecture, after changing panel to a lib panel, one can still discard the changes, so I would assume we wait with the grid item replacement until commit.

File: public/app/features/dashboard-scene/panel-edit/PanelVizTypePicker.tsx (hunk 1 above)
Comment (mdvictor) [reply public/app/features/dashboard-scene/panel-edit/PanelVizTypePicker.tsx:39]: I'll look more into this. Think we need to extend the panelModel to support the library panel prop `libraryPanel?: LibraryPanelRef | LibraryPanel;`. The `PanelLibraryOptionsGroup` component uses the PanelModel from `dashboard/state` and not the one from `grafana/data` which has these library details.

File: public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx (hunk 2 above)
Comment (mdvictor) [reply public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx:435]: What do you mean by commit? As in saving or discarding the library panel in panel edit? My thinking was that selecting a lib panel from the pane is committing and after that, if there are any other changes to said library panel you have the options to save them or discard them in the context of updating the library panel and not the context of the dashboard.

In the old arch we can discard any panel, not only library panels. But in the new arch we do not have this option for all panels, we can only go back to dashboard, thus I assumed that the discard button (and also the save) for library panels refers to changes regarding saved lib panel template.

Do you think we should update the dashboard only when the library panel is saved, after selecting one from the lib panels pane?

File: public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx [hunk 3]
```diff
@@ -409,6 +417,92 @@ export class VizPanelManager extends SceneObjectBase<VizPanelManagerState> {
...
+
+    // create the replacement libraryVizPanel
+    const libVizPanel = new LibraryVizPanel({
+      title: panel.state.title,
+      uid: libPanel.uid,
+      name: libPanel.name,
+      panelKey: panel.state.key!,
+    });
+
+    libVizPanel.activate();
```
Comment (torkelo) [inline public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx:432]: why activate here?

Comment (torkelo) [reply public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx:432]: ah, right VizPanelManager only renders the VizPanel, hm... this so messy

File: public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx (hunk 2 above)
Comment (torkelo) [reply public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx:435]: We should only update the dashboard from inside commit changes:

https://github.com/grafana/grafana/blob/main/public/app/features/dashboard-scene/panel-edit/PanelEditor.tsx#L104

**Comment by dprokop** [comment]
@mdvictor closing this one for now until we figure out a better arch for lib/vizpanels

**Pull request closed without merging**